### PR TITLE
gcc11: patch for output failure in gfortran (darwin21)

### PR DESCRIPTION
**Pull request by remkos**
#### Description

The issue that triggered this patch is that an output of a simple program like
write ( * , * ) "hello, world!"
end
comes out on the terminal, but when redirecting to a file (./a.out > file.txt) results in a zero-size file.
For the original issue:
https://fortran-lang.discourse.group/t/gfortran-code-is-not-proving-data-in-output-file-macos-monterey/2297/6
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=102992
The patch as accepted for gcc12 works fine with gcc11 as well: https://gcc.gnu.org/g:fabe8cc41e9b01913e2016861237d1d99d7567bf

The patch is to circumvent an apparent bug/limitation in the C library for MacOS 12.0 (still there in 12.0.1). Until fixed in either MacOS 12 or gcc11, we can apply the new patch.

###### Type(s)

- [x] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on
macOS 12.0.1 21A559 x86_64
Xcode Command Line Tools 13.1.0.0.1.1633545042

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- x ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by mascguy** [comment]
LGTM, at least in principle. But we need the blessing of at least one of our toolchain gurus.

@cjones051073 and @kencu, any thoughts/comments?

**Review by mascguy — APPROVED** [review]

**Comment by kencu** [comment]
I'm not really in a position to vet this at present. I think Chris will be up for it.

**Comment by remkos** [comment]
Thanks, in advance, for any reviewers to check this out. Two additional comments:
- The fix is in the runtime library, so any code already compiled with gcc11 will get the fix in runtime
- This might also be backportable to gcc10 or earlier.

**Comment by pmetzger** [comment]
@remkos does the gfortran test suite still pass after this patch is in place?

**Comment by remkos** [comment]
> @remkos does the gfortran test suite still pass after this patch is in place?

Yes, it does!

**Comment by pmetzger** [comment]
Then to the extent that I can review this (the contents of the patch are beyond my knowledge) it seems okay. I agree that we should have a couple of toolchain experts weigh in if possible.

**Comment by remkos** [comment]
I would like to point out again that upstream the same patch has already been approved for GNU CC 12, so I think it's pretty safe to apply it to our gcc11 port, as long as this issue is not fixed either upstream or in MacOS 12.

**Comment by pmetzger** [comment]
That sounds quite reasonable, and I have the expectation that this will go in very soon. I'm not myself an expert in the area, and when it comes to a port that is a commonly used prerequisite for loads of other ports, it's always nice to get another opinion. (If this were an obscure leaf of the dependency tree and not part of the trunk, I'd have merged it already.)

**Comment by kencu** [comment]
> @kencu, any thoughts/comments?

Only tht Iain is the lead gcc developer for Darwin, so as the patch comes directly from Iain, it is highly likely to be reasonable.

**Comment by remkos** [comment]
@pmetzger: yes, of course. I just wanted to add this as information to whomever would additionally review. It works nicely for me, otherwise nearly none of my Fortran programs on Monterey would provide proper output when standard output was redirected. I am just trying to help others before they start (like me) trying to revert to Big Sur because of it. 

The links provided near the top give some good analyses by the GCC experts on this issue and also provided the patch to upstream GCC 12, which is not out yet.

**Comment by remkos** [comment]
> > @kencu, any thoughts/comments?
> 
> Only that Iain is the lead gcc developer for Darwin, so as the patch comes directly from Iain, it is highly likely to be reasonable.
> 
Indeed, it comes from the highest authority 😀

**Review by kencu — COMMENTED** [review]
well done!

**Pull request merged**
